// quickmcp-sdk/dist/src/utils/logger.js
/**
 * Logger utility for QuickMCP
 * Simple logging system that respects MCP stdio constraints
 */
class Logger {
    config = {
        level: 'info',
        enabled: true,
        useStderr: true
    };
    constructor(config) {
        if (config) {
            this.config = { ...this.config, ...config };
        }
    }
    shouldLog(level) {
        if (!this.config.enabled)
            return false;
        const levels = {
            debug: 0,
            info: 1,
            warn: 2,
            error: 3
        };
        return levels[level] >= levels[this.config.level];
    }
    log(level, message, ...args) {
        if (!this.shouldLog(level))
            return;
        const timestamp = new Date().toISOString();
        const prefix = `[${timestamp}] [${level.toUpperCase()}]`;
        const fullMessage = `${prefix} ${message}`;
        // Always use stderr for MCP stdio servers to avoid corrupting JSON-RPC
        if (this.config.useStderr) {
            console.error(fullMessage, ...args);
        }
        else {
            console.log(fullMessage, ...args);
        }
    }
    debug(message, ...args) {
        this.log('debug', message, ...args);
    }
    info(message, ...args) {
        this.log('info', message, ...args);
    }
    warn(message, ...args) {
        this.log('warn', message, ...args);
    }
    error(message, ...args) {
        this.log('error', message, ...args);
    }
    configure(config) {
        this.config = { ...this.config, ...config };
    }
}
// Default logger instance
export const logger = new Logger();
// Logger factory for custom instances
export const createLogger = (config) => {
    return new Logger(config);
};
export { Logger };
