// quickmcp-sdk/dist/src/core/performance/cache-manager.js
/**
 * High-Performance Cache Manager for QuickMCP
 * Optimizes schema validation and response handling
 */
import { LRUCache } from 'lru-cache';
/**
 * Schema validation cache for performance optimization
 */
export class SchemaCache {
    zodCache = new LRUCache({
        max: 1000,
        ttl: 1000 * 60 * 60 // 1 hour
    });
    stats = {
        hits: 0,
        misses: 0
    };
    /**
     * Get cached or create Zod schema
     */
    getZodSchema(key, factory) {
        const cached = this.zodCache.get(key);
        if (cached) {
            this.stats.hits++;
            return cached;
        }
        this.stats.misses++;
        const schema = factory();
        this.zodCache.set(key, schema);
        return schema;
    }
    /**
     * Generate cache key for schema
     */
    static generateKey(schema) {
        return JSON.stringify(schema, Object.keys(schema).sort());
    }
    /**
     * Get cache statistics
     */
    getStats() {
        return {
            hits: this.stats.hits,
            misses: this.stats.misses,
            size: this.zodCache.size,
            hitRate: this.stats.hits / (this.stats.hits + this.stats.misses) || 0
        };
    }
    /**
     * Clear cache
     */
    clear() {
        this.zodCache.clear();
        this.stats.hits = 0;
        this.stats.misses = 0;
    }
}
/**
 * Response object pool for memory optimization
 */
export class ResponsePool {
    textPool = [];
    jsonPool = [];
    maxPoolSize = 100;
    /**
     * Get pooled text response object
     */
    getTextResponse() {
        return this.textPool.pop() || { type: 'text', text: '' };
    }
    /**
     * Return text response to pool
     */
    releaseTextResponse(obj) {
        if (this.textPool.length < this.maxPoolSize) {
            obj.text = '';
            this.textPool.push(obj);
        }
    }
    /**
     * Get pooled JSON response object
     */
    getJsonResponse() {
        return this.jsonPool.pop() || { type: 'text', text: '' };
    }
    /**
     * Return JSON response to pool
     */
    releaseJsonResponse(obj) {
        if (this.jsonPool.length < this.maxPoolSize) {
            obj.text = '';
            this.jsonPool.push(obj);
        }
    }
    /**
     * Get pool statistics
     */
    getStats() {
        return {
            textPool: this.textPool.length,
            jsonPool: this.jsonPool.length,
            maxSize: this.maxPoolSize
        };
    }
}
/**
 * Global cache instances
 */
export const schemaCache = new SchemaCache();
export const responsePool = new ResponsePool();
