// quickmcp-sdk/dist/src/index.js
/**
 * QuickMCP - Simplified TypeScript SDK for MCP development
 * Main entry point
 */
// Core exports
export { QuickMCPServer, createServer } from './core/server.js';
export { logger, createLogger } from './utils/logger.js';
export { Response } from './utils/response.js';
// Decorator exports
export { tool, resource, prompt, schema, getRegisteredTools, getRegisteredResources, getRegisteredPrompts, clearAllRegistries, autoRegister } from './decorators/tool.js';
// Utility exports
export { Schema, Responses, Resources, Prompts, Validate, Async, Http } from './utils/helpers.js';
/**
 * Quick server creation utility
 */
export async function quick(name) {
    const { createServer } = await import('./core/server.js');
    return createServer({ name });
}
/**
 * Create server with HTTP transport
 */
export function createHttpServer(name, port = 3000) {
    const { createServer } = require('./core/server.js');
    return createServer({
        name,
        transport: 'sse',
        http: { port }
    });
}
/**
 * Create server with STDIO transport (default)
 */
export function createStdioServer(name) {
    const { createServer } = require('./core/server.js');
    return createServer({ name, transport: 'stdio' });
}
